tests/returntypes/gradients/test_hadamard_grad.py

```diff
@@ -0,0 +1,853 @@
+                    ],
+                    [
+                        -(np.sin(x) * np.sin(y / 2) ** 2),
+                        (np.cos(x / 2) ** 2 * np.sin(y)),
+                    ],
+                    [
+                        (np.sin(x) * np.sin(y / 2) ** 2),
+                        (np.sin(x / 2) ** 2 * np.sin(y)),
+                    ],
+                    [
+                        (np.cos(y / 2) ** 2 * np.sin(x)),
+                        -(np.sin(x / 2) ** 2 * np.sin(y)),
+                    ],
+                ]
+            )
+            / 2
+        )
+        # Expvals
+        assert np.allclose(res[0][0], expval_expected[0], tol)
+        assert np.allclose(res[0][1], expval_expected[1], tol)
+
+        # Probs
+        assert np.allclose(res[1][0], probs_expected[:, 0], tol)
+        assert np.allclose(res[1][1], probs_expected[:, 1], tol)
+
+    costs_and_expected_expval = [
+        (cost1, [3]),
+        (cost2, [3]),
+        (cost3, [2, 3]),
+    ]
+
+    @pytest.mark.parametrize("cost, expected_shape", costs_and_expected_expval)
+    def test_output_shape_matches_qnode_expval(self, cost, expected_shape):
+        """Test that the transform output shape matches that of the QNode."""
+        dev = qml.device("default.qubit", wires=4)
+
+        x = np.random.rand(3)
+        circuit = qml.QNode(cost, dev)
+
+        res = qml.gradients.hadamard_grad(circuit)(x)
+        assert isinstance(res, tuple)
+        assert len(res) == expected_shape[0]
+
+        if len(expected_shape) > 1:
+            for r in res:
+                assert isinstance(r, tuple)
+                assert len(r) == expected_shape[1]
+
+    costs_and_expected_probs = [
+        (cost4, [3, 4]),
+        (cost5, [3, 4]),
+        (cost6, [2, 3, 4]),
+    ]
+
+    @pytest.mark.parametrize("cost, expected_shape", costs_and_expected_probs)
+    def test_output_shape_matches_qnode_probs(self, cost, expected_shape):
+        """Test that the transform output shape matches that of the QNode."""
+        dev = qml.device("default.qubit", wires=5)
+
+        x = np.random.rand(3)
+        circuit = qml.QNode(cost, dev)
+
+        res = qml.gradients.hadamard_grad(circuit)(x)
+        assert isinstance(res, tuple)
+        assert len(res) == expected_shape[0]
+
+        if len(expected_shape) > 2:
+            for r in res:
+                assert isinstance(r, tuple)
+                assert len(r) == expected_shape[1]
+
+                for r_ in r:
+                    assert isinstance(r_, qml.numpy.ndarray)
+                    assert len(r_) == expected_shape[2]
+
+        elif len(expected_shape) > 1:
+            for r in res:
+                assert isinstance(r, qml.numpy.ndarray)
+                assert len(r) == expected_shape[1]
+
+    def test_multi_measure_no_warning(self):
+        """Test computing the gradient of a tape that contains multiple
+        measurements omits no warnings."""
```

Comment:
I'm a bit confused what this test does. Do we need it in particular for the Hadamard test gradient?

Reply:
It is just testing that no warning are emitted, like for example ragged arrays.